### PR TITLE
Restrict member operations for hierarchical vos

### PR DESCRIPTION
**Pull request by Johaney-s**

### CODE REVIEW

**Review by xflord — COMMENTED** [review]
Looks good to me. However, do we really check all cases where member manipulation occurs? For example; I went through the moveMembership method in MembersManager and I think that in this case the original member would be deleted even if they were from a member vo. I'm not sure whether this is something we want to consider, just didn't find anything else wrong with the PR and wanted to be helpful :p

**Comment by xflord** [comment]
> > Looks good to me. However, do we really check all cases where member manipulation occurs? For example; I went through the moveMembership method in MembersManager and I think that in this case the original member would be deleted even if they were from a member vo. I'm not sure whether this is something we want to consider, just didn't find anything else wrong with the PR and wanted to be helpful :p
> 
> I was thinking about this usecase too - I think it would be correct to extend the method to move membership for both members (in parentVo and memberVo) simultanously - or we should forbid it in case it has member in parentVo OR it has member in memberVo. Do you think this would be acceptable solution? And which one would be better?

I'm not too familiar with how the method is actually used, however I feel like moving it for memberVos aswell might be confusing for admins of the VOs (new members appearing out of nowhere?). I believe the purpose of the hierarchy feature is to propagate users from members to parents, not the other way around. I would probably forbid it altogether if a VO is a part of a hierarchy. But once again, I don't know how/how often the method is used and all the other details.

**Comment by Johaney-s** [comment]
> I believe the purpose of the hierarchy feature is to propagate users from members to parents, not the other way around. 

Right, discussed with @balcirakpeter and concluded we should 1) allow to move membership for member of memberVo but also move him in parentVo, 2) allow to move membership for member of parentVo but not alter him in memberVo.

**Review by cuadradek — CHANGES_REQUESTED** [review]

File: perun-rpc/src/main/java/cz/metacentrum/perun/rpc/methods/MembersManagerMethod.java [hunk 1]
```diff
@@ -1344,6 +1346,7 @@ public Integer call(ApiCaller ac, Deserializer parms)
 	 * Deletes all VO members.
 	 *
 	 * @param vo int VO <code>id</code>
+	 * @throws MemberLifecycleAlteringForbiddenException if member comes from hierarchical member vo and cannot be altered in parent vo
```
[the remark this answers is not in the document]
Comment (Johaney-s) [reply perun-rpc/src/main/java/cz/metacentrum/perun/rpc/methods/MembersManagerMethod.java]: thanks, fixed

**Pull request merged**